File: src/wellbegun/routers/categories.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from wellbegun.database import get_db
from wellbegun.schemas.category import (
    CategoryCreate,
    CategoryOut,
    CategoryStatusCreate,
    CategoryStatusOut,
    CategoryStatusUpdate,
    CategoryUpdate,
)
from wellbegun.services import category_service

router = APIRouter(prefix="/categories", tags=["categories"])


@router.get("/", response_model=list[CategoryOut])
def list_categories(db: Session = Depends(get_db)):
    return category_service.get_all(db)


@router.get("/{category_id}", response_model=CategoryOut)
def get_category(category_id: int, db: Session = Depends(get_db)):
    category = category_service.get_by_id(db, category_id)
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    return category


@router.post("/", response_model=CategoryOut, status_code=201)
def create_category(data: CategoryCreate, db: Session = Depends(get_db)):
    return category_service.create(
        db,
        slug=data.slug,
        display_name=data.display_name,
        member_entity_type=data.member_entity_type,
    )


@router.put("/{category_id}", response_model=CategoryOut)
def update_category(
    category_id: int, data: CategoryUpdate, db: Session = Depends(get_db)
):
    updates = data.model_dump(exclude_unset=True)
    category = category_service.update(db, category_id, **updates)
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    return category


@router.delete("/{category_id}")
def delete_category(category_id: int, db: Session = Depends(get_db)):
    if not category_service.delete(db, category_id):
        raise HTTPException(status_code=404, detail="Category not found")
    return {"ok": True}


# --- Status endpoints ---

@router.get("/{category_id}/statuses", response_model=list[CategoryStatusOut])
def list_statuses(category_id: int, db: Session = Depends(get_db)):
    return category_service.get_statuses(db, category_id)


@router.post("/{category_id}/statuses", response_model=CategoryStatusOut, status_code=201)
def add_status(
    category_id: int, data: CategoryStatusCreate, db: Session = Depends(get_db)
):
    category = category_service.get_by_id(db, category_id)
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    return category_service.add_status(
        db,
        category_id=category_id,
        value=data.value,
        position=data.position,
        is_default=data.is_default,
    )


@router.put("/statuses/{status_id}", response_model=CategoryStatusOut)
def update_status(
    status_id: int, data: CategoryStatusUpdate, db: Session = Depends(get_db)
):
    updates = data.model_dump(exclude_unset=True)
    status = category_service.update_status(db, status_id, **updates)
    if not status:
        raise HTTPException(status_code=404, detail="Status not found")
    return status


@router.delete("/statuses/{status_id}")
def delete_status(status_id: int, db: Session = Depends(get_db)):
    if not category_service.delete_status(db, status_id):
        raise HTTPException(status_code=404, detail="Status not found")
    return {"ok": True}

File: frontend/src/lib/api/categories.ts
import type { Category, CategoryStatus } from '$lib/types';

const BASE = '/api/categories';

export async function getCategories(): Promise<Category[]> {
	const res = await fetch(`${BASE}/`);
	return res.json();
}

export async function getCategory(id: number): Promise<Category> {
	const res = await fetch(`${BASE}/${id}`);
	return res.json();
}

export async function createCategory(data: { slug: string; display_name: string; member_entity_type: string }): Promise<Category> {
	const res = await fetch(`${BASE}/`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	return res.json();
}

export async function updateCategory(id: number, data: Partial<Category>): Promise<Category> {
	const res = await fetch(`${BASE}/${id}`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	return res.json();
}

export async function deleteCategory(id: number): Promise<void> {
	await fetch(`${BASE}/${id}`, { method: 'DELETE' });
}

export async function getStatuses(categoryId: number): Promise<CategoryStatus[]> {
	const res = await fetch(`${BASE}/${categoryId}/statuses`);
	return res.json();
}

export async function addStatus(categoryId: number, data: { value: string; position?: number; is_default?: boolean }): Promise<CategoryStatus> {
	const res = await fetch(`${BASE}/${categoryId}/statuses`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	return res.json();
}

export async function updateStatus(statusId: number, data: { value?: string; position?: number; is_default?: boolean }): Promise<CategoryStatus> {
	const res = await fetch(`${BASE}/statuses/${statusId}`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data)
	});
	return res.json();
}

export async function deleteStatus(statusId: number): Promise<void> {
	await fetch(`${BASE}/statuses/${statusId}`, { method: 'DELETE' });
}
